src/coingecko/Coingecko.ts

```diff
@@ -82,30 +145,74 @@ export class Coingecko {
     currency = "usd",
     platform_id = "ethereum"
   ): Promise<CoinGeckoPrice[]> {
-    // Generate a unique set with no repeated. join the set with the required coingecko delimiter.
-    const contract_addresses = Array.from(new Set(addresses.filter((n) => n).values()));
+    const priceCache: { [addr: string]: CoinGeckoPrice } = this.getPriceCache(currency, platform_id);
+
+    // Pre-populate price cache with requested token addresses
+    addresses.forEach((addr: string) => {
+      if (priceCache[addr] === undefined) {
+        priceCache[addr] = { address: addr, price: 0, timestamp: 0 };
+      }
+    });
+
+    // Collect all known token addresses (requested + other cached).
+    const contract_addresses: string[] = Object.keys(priceCache);
     assert(contract_addresses.length > 0, "Must supply at least 1 contract address");
-    // coingecko returns lowercase addresses, so if you expect checksummed addresses, this lookup table will convert them back without having to add ethers as a dependency
-    const lookup = Object.fromEntries(
-      contract_addresses.map((address) => {
-        return [address.toLowerCase(), address];
-      })
-    );
+    this.logger.debug({
+      at: "Coingecko#getContractPrices",
+      message: `Updating ${platform_id}/${currency} token prices.`,
+      tokens: contract_addresses,
+    });
+
     // annoying, but have to type this to iterate over entries
+    type CGTokenPrice = {
+      [currency: string]: number;
+      last_updated_at: number;
+    };
     type Result = {
-      [address: string]: {
-        usd: number;
-        last_updated_at: number;
-      };
+      [address: string]: CGTokenPrice;
     };
+    // Coingecko expects a comma-delimited (%2c) list.
     const result: Result = await this.call(
       `simple/token_price/${platform_id}?contract_addresses=${contract_addresses.join(
         "%2C"
       )}&vs_currencies=${currency}&include_last_updated_at=true`
     );
-    return Object.entries(result).map(([key, value]) => {
-      return { address: lookup[key], timestamp: value.last_updated_at, price: value.usd };
+
+    // Note: contract_addresses is a reliable reference for the price lookup.
+    // priceCache might have been updated subsequently by concurrent price requests.
+    const updated: string[] = [];
+    contract_addresses.forEach((addr) => {
+      const cgPrice: CGTokenPrice | undefined = result[addr.toLowerCase()];
+
+      if (cgPrice === undefined) {
+        this.logger.debug({
+          at: "Coingecko#getContractPrices",
+          message: `Token ${addr} not included in CoinGecko response.`,
+        });
+      } else if (cgPrice.last_updated_at > priceCache[addr].timestamp) {
+        priceCache[addr] = {
```

Comment:
something like cache.set(key,data,timestamp) would be nice here and only sets if timestamp is newer